File: c/vec3.c
#include <math.h>
#include <stdio.h>
#include <stdlib.h>

#include "vec3.h"

vec3 vec3_new(float x, float y, float z) {
    vec3 v;
    v.x = x;
    v.y = y;
    v.z = z;
    return v;
}

float vec3_length(vec3 v) {
    return sqrt(pow(v.x, 2) + pow(v.y, 2) + pow(v.z, 2));
}

vec3 vec3_scale(vec3 v, float scale) {
    vec3 result;
    result.x = v.x * scale;
    result.y = v.y * scale;
    result.z = v.z * scale;
    return result;
}

vec3 vec3_normalize(vec3 v) {
    // do degenerate case check
    if (vec3_length(v) == 0) {
        return v;
    }

    // scale by 1 / length, which sends length(v) to 1
    return vec3_scale(v, 1.0 / vec3_length(v));
}

vec3 vec3_add(vec3 v1, vec3 v2) {
    vec3 result;
    result.x = v1.x + v2.x;
    result.y = v1.y + v2.y;
    result.z = v1.z + v2.z;
    return result;
}

vec3 vec3_sub(vec3 v1, vec3 v2) {
    // negate v2, store into result, and then add the negation
    return vec3_add(v1, vec3_scale(v2, -1.0));
}

float vec3_dot(vec3 v1, vec3 v2) {
    return (v1.x * v2.x) + (v1.y * v2.y) + (v1.z * v2.z);
}

vec3 vec3_random_in_unit_sphere() {
    vec3 v;
    while (1) {
        // vec3 with components a random length from 0 to 1
        vec3 random_vec3 = vec3_new(drand48(), drand48(), drand48());
        // double it; components random from 0 to 2
        random_vec3 = vec3_scale(random_vec3, 2);
        // this is just a vec3 with all 1 components. Bad name. Whatever.
        vec3 cube_diag = vec3_new(1, 1, 1);
        // we will create a vec3 that lands somewhere in the -1 to +1 cube
        vec3 unit_cube_vec3 = vec3_sub(cube_diag, random_vec3);

        // check if unit_cube_vec3 is longer than 1; AKA, outside of the unit sphere.
        if (vec3_length(unit_cube_vec3) < 1.0) {
            v = unit_cube_vec3;
            break;
        }
    }
    return v;
}

vec3 vec3_gamma_correct(vec3 color, float gamma) {
    vec3 result;
    result.x = pow(color.x, gamma);
    result.y = pow(color.y, gamma);
    result.z = pow(color.z, gamma);
    return result;
}

vec3 vec3_reflect(vec3 v, vec3 normal) {
    // return v - 2 * dot(v, n) * n
    float two_dot_VN = 2 * vec3_dot(v, normal);
    vec3 scaled_normal = vec3_scale(normal, 2 * two_dot_VN);
    return vec3_sub(v, scaled_normal);
}

File: java/src/raytracer/Vec3.java
package raytracer;

import java.util.Random;

/**
 * A simple 3D vector class, with the operations necessary for use in the
 * pathtracer.
 * 
 * @author nick
 *
 */
public class Vec3 {
	private double x, y, z;

	/**
	 * Creates a new instance of a Vec3 with the given components.
	 * 
	 * @param x the x component of the new Vec3.
	 * @param y the y component of the new Vec3.
	 * @param z the z component of the new Vec3.
	 */
	public Vec3(double x, double y, double z) {
		this.x = x;
		this.y = y;
		this.z = z;
	}

	/**
	 * Calculates the length of this Vec3.
	 * 
	 * @return the length of this Vec3.
	 */
	public double length() {
		return Math.sqrt(Math.pow(x, 2) + Math.pow(y, 2) + Math.pow(z, 2));
	}

	/**
	 * Multiplies each component of this Vec3 by a given scalar.
	 * 
	 * @param v      the Vec3 to scale.
	 * @param scalar the scalar value to multiply.
	 * @return a new Vec3 with values of the result of the scaling.
	 */
	public static Vec3 scale(Vec3 v, double scalar) {
		Vec3 result = new Vec3(0, 0, 0);
		result.x = v.x * scalar;
		result.y = v.y * scalar;
		result.z = v.z * scalar;
		return result;
	}

	/**
	 * Scales this Vec3 so that its length is 1.0. If <code>v</code> has a length of
	 * 0, the original Vec3 is returned.
	 * 
	 * @param v the Vec3 to normalize.
	 * @return a new Vec3 with values of the normalized original Vec3.
	 */
	public static Vec3 normalize(Vec3 v) {
		if (v.length() == 0) {
			return v;
		}

		return Vec3.scale(v, 1 / v.length());
	}

	/**
	 * Sums two Vec3.
	 * 
	 * @param left  the left operand of the subtraction.
	 * @param right the right operand of the subtraction.
	 * @return the sum of the given Vec3.
	 */
	public static Vec3 add(Vec3 left, Vec3 right) {
		Vec3 result = new Vec3(0, 0, 0);

		result.setX(left.getX() + right.getX());
		result.setY(left.getY() + right.getY());
		result.setZ(left.getZ() + right.getZ());

		return result;
	}

	/**
	 * Subtracts two Vec3.
	 * 
	 * @param left  the left operand of the subtraction.
	 * @param right the right operand of the subtraction.
	 * @return the difference of the given Vec3.
	 */
	public static Vec3 sub(Vec3 left, Vec3 right) {
		Vec3 result = new Vec3(0, 0, 0);

		result.setX(left.getX() - right.getX());
		result.setY(left.getY() - right.getY());
		result.setZ(left.getZ() - right.getZ());

		return result;
	}

	/**
	 * Calculates the dot product of two Vec3.
	 * 
	 * @param left  the left operand of the dot product.
	 * @param right the right operand of the dot product.
	 * @return the dot product of the given Vec3.
	 */
	public static double dot(Vec3 left, Vec3 right) {
		return (left.getX() * right.getX()) + (left.getY() * right.getY()) + (left.getZ() * right.getZ());
	}

	@Override
	public boolean equals(Object o) {
		if (o == null) {
			return false;
		}

		if (!(o instanceof Vec3)) {
			return false;
		}

		Vec3 other = (Vec3) o;

		return (getX() == other.getX()) && (getY() == other.getY()) && (getZ() == other.getZ());
	}

	public double getX() {
		return x;
	}

	public void setX(double x) {
		this.x = x;
	}

	public double getY() {
		return y;
	}

	public void setY(double y) {
		this.y = y;
	}

	public double getZ() {
		return z;
	}

	public void setZ(double z) {
		this.z = z;
	}

	/**
	 * Returns a random vec3 with a length of 1.0 or less. Each component of the
	 * vec3 is in the range [-1, 1], as long as the length of the vec3 is 1.0 or
	 * less.
	 * 
	 * @return a random vec3 with a length of 1.0 or less.
	 */
	public static Vec3 randomInUnitSphere() {
		Random random = new Random();
		while (true) {
			// vec3 with components a random length from 0 to +1
			Vec3 randomVec3 = new Vec3(random.nextDouble(), random.nextDouble(), random.nextDouble());
			// double it; components random from 0 to +2
			randomVec3 = Vec3.scale(randomVec3, 2);
			// this is just a vec3 with all 1 components. Bad name. Whatever.
			Vec3 cubeDiag = new Vec3(1, 1, 1);
			// we will create a vec3 that lands somewhere in the -1 to +1 cube
			Vec3 unitCubeVec3 = Vec3.sub(cubeDiag, randomVec3);

			// check if unit_cube_vec3 is longer than 1; AKA, outside of the unit sphere.
			if (unitCubeVec3.length() < 1.0) {
				return unitCubeVec3;
			}
		}
	}

	/**
	 * Returns a new Vec3 that is the value of a Vec3 reflecting off a surface with
	 * the given normal.
	 * 
	 * @param v      the incoming Vec3
	 * @param normal the normal to reflect <code>v</code> off of
	 * @return a new Vec3 that is the reflection of the incoming Vec3
	 */
	public static Vec3 reflect(Vec3 v, Vec3 normal) {
		// return v - 2 * dot(v, n) * n
		double twoDotVN = 2 * Vec3.dot(v, normal);
		return Vec3.sub(v, Vec3.scale(normal, 2 * twoDotVN));
	}

	/**
	 * Gamma corrects this Vec3, under the interpretation that it is representing a
	 * color.
	 * 
	 * @param color the input color to gamma correct.
	 * @param gamma the value to gamma correct by, which is the power to take all
	 *              components to.
	 * @return a new Vec3 as the result of the gamma correction.
	 */
	public static Vec3 gammaCorrect(Vec3 color, double gamma) {
		Vec3 result = new Vec3(0, 0, 0);
		result.x = Math.pow(color.getX(), gamma);
		result.y = Math.pow(color.getY(), gamma);
		result.z = Math.pow(color.getZ(), gamma);
		return result;
	}

	/**
	 * {@inheritDoc}
	 */
	@Override
	public String toString() {
		return String.format("Vec3(%f, %f, %f)", x, y, z);
	}
}
